number of bug fixes and improvements

piru opened the pull request:
I've added more error checking that was a bit lacking and fixed couple of bugs such as memory leaks and a double free in error code path.

The changes in detail:

- utilize do {} while(0) -construct for ERR and AV_ERR macros to avoid excess ;
- use ERR and AV_ERR macros when fitting
- make local functions static
- ac_init: added missing error checks
- ac_probe_input_buffer: added missing error checks, only memset padding area
- ac_probe_input_stream: added missing error checks, removed unnecessary memset, fixed a double free, fixed a memory leak
- cpymetadata: fixed to always '\0' terminate the result buffer
- ac_open: added missing error checks
- ac_get_stream_info: added missing error checks
- ac_create_video_decoder: added missing error checks
- ac_create_audio_decoder: added missing error checks
- ac_decode_video_package: added missing error checks
- ac_decode_audio_package: don't leak memory if av_realloc fails
- ac_free_audio_decoder: don't leak pDecoder->pCodecCtx

astoeckel commented:
Thank you for this quite comprehensive PR! Acinerella was the first serious C program I ever wrote back in high school, so that is where a lot of the issues you've fixed come from ;-)

One request: could you separate the part that adds the additional callback to `ac_create_decoder` into another PR and add some more rationale for why this is needed? I'm not entirely sure whether it is a good idea to leak the FFmpeg datastrucutres through the Acinerella API; however, I would be fine with a function `ac_create_decoder_ex` that does that.

Commit:
- utilize do {} while(0) -construct for ERR and AV_ERR macros to avoid excess ;
- use ERR and AV_ERR macros when fitting
- make local functions static
- ac_init: added missing error checks
- ac_probe_input_buffer: added missing error checks, only memset padding area
- ac_probe_input_stream: added missing error checks, removed unnecessary memset, fixed a double free, fixed a memory leak
- cpymetadata: fixed to always '\0' terminate the result buffer
- ac_open: added missing error checks
- ac_get_stream_info: added missing error checks
- ac_create_video_decoder: added missing error checks
- ac_create_audio_decoder: added missing error checks
- ac_decode_video_package: added missing error checks
- ac_decode_audio_package: don't leak memory if av_realloc fails
- ac_free_audio_decoder: don't leak pDecoder->pCodecCtx

piru commented:
Sounds sensible. Here's the edited first PR with just the fixes and no API changes.

astoeckel commented:
Tested; doesn't seem to introduce any regressions. ;-)

Thank you!
